File: day5/cxx/puzzle2.cxx
#include <algorithm>
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

#include "util.hpp"

using namespace std;

const auto solve()
{
    ifstream f{"../input.txt"};

    auto parts = get_parts(f);

    auto stacks = build_initial_stacks(std::get<0>(parts));

    for (auto command: std::get<1>(parts))
    {
        execute_9001_command(stacks, command);
    }

    auto top_crates = ""s;

    for (auto stack: stacks)
    {
        top_crates += stack.back();
    }

    return top_crates;
}

int main()
{
    auto solution = solve();
    cout << solution << endl;
}

File: day5/py/puzzle2.py
#!/usr/bin/env python

from util import get_parts, build_initial_stacks, execute_9001_command

def solve():
    with open('../input.txt', 'r') as f:
        (initial_stacks, commands) = get_parts(f)

        stacks = build_initial_stacks(initial_stacks)

        for command in commands:
            stacks = execute_9001_command(stacks, command)

    top_crates = ''

    for stack in stacks:
        top_crates += stack[-1]

    return top_crates

if __name__ == "__main__":
    solution = solve()
    print(solution, )

